File: java/Solution78.java
package leetCode.java;

import java.util.*;

//Backtracking O(n * 2^n) O(n)
class Solution78 {
    List<List<Integer>> res = new ArrayList<>();

    public List<List<Integer>> subsets(int[] nums) {
        backtracking(0, new ArrayList<>(), nums);
        return res;    
    }

    public void backtracking(int start, List<Integer> selected, int[] nums) {
        res.add(new ArrayList<>(selected)); // O(n), 記得建立新物件

        for (int i = start; i < nums.length; ++i) {
            selected.add(nums[i]);
            backtracking(i + 1, selected, nums);
            selected.remove(selected.size() - 1);
        }
    }
}
/*
nums = [1,2,3]
                        []
        1               2               3
    12     13           23  
    123  
*/

File: c++/Solution78.cpp
// Backtracking O(n * 2^n) O(n)
class Solution78{
public:
    vector<vector<int>> res;

    vector<vector<int>> subsets(vector<int>& nums) {
        vector<int> selected;
        backtracking(0, selected, nums); // 不用sort因為沒有重複的數字
        return res;
    }

    void backtracking(int start, vector<int>& selected, vector<int>& nums) {
        res.push_back(selected);

        for (int i = start; i < nums.size(); ++i) {
            selected.push_back(nums[i]);
            backtracking(i + 1, selected, nums);
            selected.pop_back();
        }
    }
};